Fix TestTestingFarmUtil.test_fetch_failed_test_cases_from_file

The above named test no longer works. It looks like it tried to fetch a file from testing-farm and potentially this file does no longer exist. I suggest to find a different failing testing-farm request and store the output locally. Then through dependency injection try to fake the call to testing-farm to fetch the file at any given time from disk instead the web.

Here's the log of the failing test:

```
=================================== FAILURES ===================================
__________ TestTestingFarmUtil.test_fetch_failed_test_cases_from_file __________

self = <tests.testing_farm_util_test.TestTestingFarmUtil testMethod=test_fetch_failed_test_cases_from_file>

    def test_fetch_failed_test_cases_from_file(self):
        request_id = "1f25b0df-71f1-4a13-a4b8-c066f6f5f116"
        chroot = "fedora-39-x86_64"
        req = tf.TestingFarmRequest(
            request_id=request_id,
            chroot=chroot,
            copr_build_ids=[11, 22, 33],
        )
        actual = req.get_failed_test_cases_from_xunit_file(
            xunit_file=self.abspath(f"testing-farm-logs/results_{request_id}.xml"),
            artifacts_url_origin="https://example.com/",
        )
    
        expected = [
            tf.FailedTestCase(
                test_name="/compiler-rt-tests/cross-compile-i686",
                request_id=request_id,
                chroot=chroot,
                log_output_url=f"[https://artifacts.dev.testing-farm.io/{request_id}/work-snapshot-gatingk5n38qtr/tests/snapshot-gating/execute/data/guest/default-0/compiler-rt-tests/cross-compile-i686-18/output.txt](https://artifacts.dev.testing-farm.io/%7Brequest_id%7D/work-snapshot-gatingk5n38qtr/tests/snapshot-gating/execute/data/guest/default-0/compiler-rt-tests/cross-compile-i686-18/output.txt)",
                log_output="+ clang -m32 -fsanitize=address test.c\n"
                "/usr/bin/ld: cannot find -lgcc_s: No such file or "
                "directory\n"
                "clang: error: linker command failed with exit code "
                "1 (use -v to see invocation)\n"
                "Shared connection to 3.12.104.11 closed.\n",
                artifacts_url="https://example.com/",
            )
        ]
>       self.assertEqual(actual, expected)
E       AssertionError: Lists differ: [Fail[371 chars]put='<html>\n<head><title>404 Not Found</title[147 chars]om')] != [Fail[371 chars]put='+ clang -m32 -fsanitize=address test.c\n/[218 chars]om')]
E       
E       First differing element 0:
E       Faile[370 chars]put='<html>\n<head><title>404 Not Found</title[146 chars]com')
E       Faile[370 chars]put='+ clang -m32 -fsanitize=address test.c\n/[217 chars]com')
E       
E         [FailedTestCase(test_name='/compiler-rt-tests/cross-compile-i686',
E                         request_id='1f25b0df-71f1-4a13-a4b8-c066f6f5f116',
E                         chroot='fedora-39-x86_64',
E                         log_output_url='https://artifacts.dev.testing-farm.io/1f25b0df-71f1-4a13-a4b8-c066f6f5f116/work-snapshot-gatingk5n38qtr/tests/snapshot-gating/execute/data/guest/default-0/compiler-rt-tests/cross-compile-i686-18/output.txt',
E       -                 log_output='<html>\n'
E       -                            '<head><title>404 Not Found</title></head>\n'
E       +                 log_output='+ clang -m32 -fsanitize=address test.c\n'
E       +                            '/usr/bin/ld: cannot find -lgcc_s: No such file or '
E       -                            '<body>\n'
E       ?                             ^^ ^ -
E       
E       +                            'directory\n'
E       ?                             ^^^^^^ ^
E       
E       +                            'clang: error: linker command failed with exit code '
E       +                            '1 (use -v to see invocation)\n'
E       +                            'Shared connection to 3.12.104.11 closed.\n',
E       -                            '<center><h1>404 Not Found</h1></center>\n'
E       -                            '<hr><center>nginx/1.20.0</center>\n'
E       -                            '</body>\n'
E       -                            '</html>\n',
E                         artifacts_url='https://example.com/')]

snapshot_manager/tests/testing_farm_util_test.py:37: AssertionError
------------------------------ Captured log call -------------------------------
INFO     root:util.py:133 Getting URL https://artifacts.dev.testing-farm.io/1f25b0df-71f1-4a13-a4b8-c066f6f5f116/work-snapshot-gatingk5n38qtr/tests/snapshot-gating/execute/data/guest/default-0/compiler-rt-tests/cross-compile-i686-18/output.txt
```

--- a/snapshot_manager/snapshot_manager/testing_farm_util.py
+++ b/snapshot_manager/snapshot_manager/testing_farm_util.py
@@ -42,6 +42,12 @@ class TestingFarmRequest:
     )
     copr_build_ids_pattern: ClassVar[str] = r"(\d,)*\d"
 
+    _in_test_mode: bool = False
+    """When this mode is on, we can workaround certain restrictions for fetching
+    outdated URLs for example."""
+
+    _dirname: str = pathlib.Path(os.path.dirname(__file__))
+
     @property
     def are_build_ids_still_valid(self, copr_build_ids: list[int]) -> bool:
         """Returns True if the given copr builds are the same as the ones
@@ -110,14 +116,30 @@ class TestingFarmRequest:
         >>> keys = requests.keys()
         >>> keys
         dict_keys(['fedora-rawhide-x86_64', 'fedora-39-x86_64', 'fedora-40-x86_64', 'fedora-38-x86_64'])
-        >>> requests['fedora-rawhide-x86_64']
-        TestingFarmRequest(request_id=UUID('271a79e8-fc9a-4e1d-95fe-567cc9d62ad4'), chroot='fedora-rawhide-x86_64', copr_build_ids=[1, 2, 3])
-        >>> requests['fedora-39-x86_64']
-        TestingFarmRequest(request_id=UUID('22222222-fc9a-4e1d-95fe-567cc9d62ad4'), chroot='fedora-39-x86_64', copr_build_ids=[5, 6, 7])
-        >>> requests['fedora-40-x86_64']
-        TestingFarmRequest(request_id=UUID('33333333-fc9a-4e1d-95fe-567cc9d62ad4'), chroot='fedora-40-x86_64', copr_build_ids=[12, 13, 14])
-        >>> requests['fedora-38-x86_64']
-        TestingFarmRequest(request_id=UUID('44444444-fc9a-4e1d-95fe-567cc9d62ad4'), chroot='fedora-38-x86_64', copr_build_ids=[])
+        >>> requests['fedora-rawhide-x86_64'].request_id
+        UUID('271a79e8-fc9a-4e1d-95fe-567cc9d62ad4')
+        >>> requests['fedora-rawhide-x86_64'].chroot
+        'fedora-rawhide-x86_64'
+        >>> requests['fedora-rawhide-x86_64'].copr_build_ids
+        [1, 2, 3]
+        >>> requests['fedora-39-x86_64'].request_id
+        UUID('22222222-fc9a-4e1d-95fe-567cc9d62ad4')
+        >>> requests['fedora-39-x86_64'].chroot
+        'fedora-39-x86_64'
+        >>> requests['fedora-39-x86_64'].copr_build_ids
+        [5, 6, 7]
+        >>> requests['fedora-40-x86_64'].request_id
+        UUID('33333333-fc9a-4e1d-95fe-567cc9d62ad4')
+        >>> requests['fedora-40-x86_64'].chroot
+        'fedora-40-x86_64'
+        >>> requests['fedora-40-x86_64'].copr_build_ids
+        [12, 13, 14]
+        >>> requests['fedora-38-x86_64'].request_id
+        UUID('44444444-fc9a-4e1d-95fe-567cc9d62ad4')
+        >>> requests['fedora-38-x86_64'].chroot
+        'fedora-38-x86_64'
+        >>> requests['fedora-38-x86_64'].copr_build_ids
+        []
         """
         matches = re.findall(r"<!--TESTING_FARM:([^/]+)/([^/]+)(/([^/]+))?-->", string)
         if not matches:
@@ -453,7 +475,14 @@ class TestingFarmRequest:
                 './logs/log[@name="testout.log"]'
             ).get("href")
 
-            log_file = util.read_url_response_into_file(log_output_url)
+            log_file: pathlib.Path
+            if not self._in_test_mode:
+                log_file = util.read_url_response_into_file(log_output_url)
+            else:
+                p = self._dirname.joinpath(
+                    f"../tests/testing-farm-logs/output_{self.request_id}.txt"
+                )
+                log_file = pathlib.Path(p)
             tc = FailedTestCase(
                 test_name=failed_testcase.get("name"),
                 log_output_url=log_output_url,

--- a/snapshot_manager/tests/testing-farm-logs/output_1f25b0df-71f1-4a13-a4b8-c066f6f5f116.txt
+++ b/snapshot_manager/tests/testing-farm-logs/output_1f25b0df-71f1-4a13-a4b8-c066f6f5f116.txt
@@ -1,0 +1,4 @@
++ clang -m32 -fsanitize=address test.c
+/usr/bin/ld: cannot find -lgcc_s: No such file or directory
+clang: error: linker command failed with exit code 1 (use -v to see invocation)
+Shared connection to 3.12.104.11 closed.

--- a/snapshot_manager/tests/testing_farm_util_test.py
+++ b/snapshot_manager/tests/testing_farm_util_test.py
@@ -13,6 +13,7 @@ class TestTestingFarmUtil(base_test.TestBase):
             request_id=request_id,
             chroot=chroot,
             copr_build_ids=[11, 22, 33],
+            _in_test_mode=True,
         )
         actual = req.get_failed_test_cases_from_xunit_file(
             xunit_file=self.abspath(f"testing-farm-logs/results_{request_id}.xml"),

